### PR TITLE
Desktop #2383: Copy To Clipboard 

### PR DESCRIPTION
**Pull request by Runo-saduwa**
### SCREENSHOT SHOWING THE BUTTON STYLE
copy-paste-joplin

### TEST CASES
- Try with a code block - button appears
- Try with an inline code block - no button
- Try with a Fountain block - no button
- Try with a Katex block - no button
- Try with Katex inline code - no button
- Try with unformatted text (indented) that contains code - no button

### CODE REVIEW

**Comment by laurent22** [comment]
Thanks for the PR. Please install the precommit hook as described in BUILD.md as there are formatting issues.

I didn't have time to answer your questions in the original issue, but this feature would need to be developed as a markdown-it plugin. It allows the code to be more maintainable, by encapsulating the feature, and it means it's also less likely to interfere with other plugins.

It also mean that by developing it once it will work on both desktop on mobile.

So please could you repackage it as a plugin? You'll find some examples in the joplin-renderer directory, and you might also want to read the markdown-it doc to understand how the plugins work.

**Comment by Runo-saduwa** [comment]
I will work on it again, thanks for the information!💪🏾

**Comment by Runo-saduwa** [comment]
Ok here come again, how do I create a plugin?, I’ve learnt how to add a
plugin in to Markdown-it, but making plugin is the headache, I’ve looked
through the MdToHtml/rules directory which seems to be where these plugins
are created , but it’s not clear how I’ll build this feature

**Comment by laurent22** [comment]
Have a look for example at the code_inline rule - it is used to add a class to a CODE element. You could duplicate this rule and instead process the code_block tokens. As a test, for example see if you can add a class to such a token.

Once you get this basic thing working, try to insert HTML elements within a code block (see the other rules to see how it's done). This is the key - once you managed to insert these extra elements, you can insert your button and position it the way you want with CSS.

See the "checkbox" rule to see how to add CSS from a rule.

But also check the markdown-it developer doc. It will tell you many details about how rules work: https://github.com/markdown-it/markdown-it/tree/master/docs

**Comment by Runo-saduwa** [comment]
Thanks, I’ll do my best and get back to you as soon as possible

**Comment by Runo-saduwa** [comment]
Hi @laurent22 , i am back again. i tried to follow your guide as much as possible, but some things still weren't quite clear, even the markdown-it docs is really unfriendly and i've done a lot of google but it seems like very few people have done stuff like this. but im really determined to solve this issue.

i have tried to implement the instructions you gave in the first paragraph, please kindly review and help me out with what needs to be changed and added next.

**Comment by laurent22** [comment]
That seems about right. You need to enable your rule in MdToHtml and then verify that it adds your class using the code inspector (in Help menu).

**Comment by Runo-saduwa** [comment]
Alright, let me confirm that

**Comment by Runo-saduwa** [comment]
@laurent22 i see it, Lol

i guess the next thing is to insert, elements inside the code block, can you please show me a really concrete and clear example to get this one done

**Comment by laurent22** [comment]
Best is to read through the markdown-it document again as this is where you'll find the most basic examples: https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md

Maybe also check our /rules directory, in particular the checkbox rule (but it's more complex).

**Comment by Runo-saduwa** [comment]
work in progress, i will give updates on progress so far, its going fine. Thanks for the guidance

**Comment by Runo-saduwa** [comment]
i have successfully developed the plugin.  I will still have to polish the styling but i will like the maintainers to test this branch before i finalize the styling. thank you

**Comment by Runo-saduwa** [comment]
i have completed the styling, the PR is ready for review now

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/copyFunction.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+(function(btn) {
+	const MDIC = {
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/copyFunction.js]: Please use full words for function names and variables. `btn` = `button`. `MDIC` I don't know what it means.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
+const fs = require('fs');
+const path = require('path');
+const filePath = path.join(__dirname, '../copyFunction.js');
+const btnCode = fs.readFileSync(filePath).toString();
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: That's an interesting approach to load the JS, but that won't work here because the "fs" package or reading file from the filesystem won't work on mobile (or browser, where this package can also be used). I think you can simply bring bring the copyFunction in this file directly.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+
+const copyCodeStyle = `
+.hljs {
+	position: relative;
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Why does the hljs class need to be changed? If it does need to be changed, it should in noteStyle.js directly.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+const LAST_TAG = '</pre>';
+
+const enhance = (render, options = {}) => (...args) => {
+	/* args = [tokens, idx, options, env, slf] */
+	const {
+		failText = 'copy fail', // copy-fail text
+		successText = 'copied!', // copy-success text
+		successTextDelay = 2000, // successText show time [ms]
+	} = options;
+	const [tokens = {}, idx = 0] = args;
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: No default value please as the value should be there. If they aren't, you can throw an error (but probably not needed).

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		`data-mdic-notify-id="${uuid}" `,
+		`data-mdic-notify-delay="${successTextDelay}" `,
+		`data-mdic-copy-fail-text="${failText}" `,
+		`onclick="${btnCode}"><i class="fa fa-copy"></i></button>`,
+		'</div>',
+	];
+	const copyTag = buttonBuilder.join('');
+
+	const originResult = render.apply(this, args);
+	const newResult = originResult.replace(LAST_TAG, `${copyTag}${LAST_TAG}`);
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: We shouldn't assume that the code block ends with a </PRE>. What if it's changed to a DIV later, or what if there are multiple </PRE> inside the block? (which I think can happen) Is there a more reliable way to insert this content without making assumption about how it ends?

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+const enhance = (render, options = {}) => (...args) => {
+	/* args = [tokens, idx, options, env, slf] */
+	const {
+		failText = 'copy fail', // copy-fail text
+		successText = 'copied!', // copy-success text
+		successTextDelay = 2000, // successText show time [ms]
+	} = options;
+	const [tokens = {}, idx = 0] = args;
+	// get content
+	const cont = tokens[idx].content;
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: `cont` = `content`

**Comment by laurent22** [comment]
Thanks @Runo-saduwa, we're getting there and your approach seems correct overall. Did you try on the desktop app and does it work as expected?

I think we'll need to test the following cases on desktop:

- Try with a code block - button should appear
- Try with an inline code block - no button
- Try with a Fountain block - no button
- Try with a Katex block - no button
- Try with Katex inline code - no button
- Try with unformatted text (indented) that contains code - no button

See here for how to create these examples: https://joplinapp.org/markdown/

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 6 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: done

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 5 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: i'll look into this

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 4 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: defaults removed

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 3 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: i had to give it a ```position: relative```, to be able to to position the copy button with ```position:absolute```, but i'll take the style to noteStyle.js

Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: change completed

**Comment by laurent22** [comment]
Also please attach a screenshot to show how the button looks.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 2 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: done

**Comment by Runo-saduwa** [comment]
Good day @laurent22 , please can you suggest a way I could insert the
button into the <pre> element. This is my only challenge now

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 5 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Good day @laurent22 , please can you suggest a way I could insert the
button into the <pre> element. This is my only challenge now

**Comment by laurent22** [comment]
> Good day @laurent22 , please can you suggest a way I could insert the
> button into the <pre> element. This is my only challenge now

I don't know actually, as I haven't developed this type of plugin yet. You might need to experiment and try various things based on the documentation or existing examples.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml.js [hunk 7]
```diff
@@ -11,12 +18,28 @@ const rules = {
...
+>>>>>>> loaded copy-to-clipboard rule
+=======
+	copy: require('./MdToHtml/rules/copy_to_clipboard'),
+>>>>>>> added copy rule to MdToHtml.js
+=======
+>>>>>>> clean up
+=======
+	copy: require('./MdToHtml/rules/copy_to_clipboard'),
+>>>>>>> finished up css
 };
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: That won't build.

File: ReactNativeClient/lib/joplin-renderer/assets/highlight.js/atom-one-light.css [hunk 8]
```diff
@@ -26,6 +26,7 @@ hue-6-2: #c18401
   background: #fafafa;
 }
 
+
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/assets/highlight.js/atom-one-light.css]: Please remove this file from the commit.

File: package.json [hunk 9]
```diff
@@ -36,5 +36,8 @@
     "husky": "^3.0.2",
     "lint-staged": "^9.2.1",
     "typescript": "^3.7.3"
+  },
+  "dependencies": {
+    "tesseract.js": "^2.0.2"
```
Comment (laurent22) [inline package.json]: Unrelated to commit

File: package-lock.json [hunk 10]
```diff
@@ -364,6 +364,11 @@
       "integrity": "sha1-ibTRmasr7kneFk6gK4nORi1xt2c=",
```
Comment (laurent22) [inline package-lock.json]: Please remove this file from commit

File: ReactNativeClient/lib/joplin-renderer/MdToHtml.js [hunk 11]
```diff
@@ -2,6 +2,13 @@ const MarkdownIt = require('markdown-it');
 const md5 = require('md5');
 const noteStyle = require('./noteStyle');
 const { fileExtension } = require('./pathUtils');
+
+const optionsCopy = {
+	failText: 'copy fail', // 'copy fail' | copy-fail text
+	successText: 'copied to clipboard!', // 'copy success' | copy-success text
+	successTextDelay: '2000', // 2000 | successText show time [ms]
+};
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: This structure appears in two locations, but you don't even need it. Please put the text and delay directly where it's needed.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml.js [hunk 12]
```diff
@@ -161,7 +184,6 @@ class MdToHtml {
 					context.pluginAssets['highlight.js'] = [
 						{ name: options.codeTheme },
 					];
-
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: The commit would be much easier to read if it wasn't for all these unnecessary white space changes.

**Comment by laurent22** [comment]
Sorry too many problems with the commit at the moment, so I can't review. Please remove the unnecessary white space changes; remove the code unrelated to the PR, and please fix the multiple conflicts.

**Comment by Runo-saduwa** [comment]
@laurent22 Thanks for the last review, i have made all the changes you asked. kindly review.

**Comment by tessus** [comment]
The file `ReactNativeClient/lib/joplin-renderer/assets/highlight.js/atom-one-light.css` is still in your PR.

**Comment by Runo-saduwa** [comment]
@tessus thanks for reminding me, i actually thought i got rid of that.

**Comment by Runo-saduwa** [comment]
@laurent22 Good day, i’m waiting for feedback

File: ReactNativeClient/lib/joplin-renderer/MdToHtml.js [hunk 13]
```diff
@@ -11,6 +11,7 @@ const rules = {
 	highlight_keywords: require('./MdToHtml/rules/highlight_keywords'),
 	code_inline: require('./MdToHtml/rules/code_inline'),
 	fountain: require('./MdToHtml/rules/fountain'),
+	copy: require('./MdToHtml/rules/copy_to_clipboard'),
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: copy => copy_to_clipboard

File: ElectronClient/app/gui/note-viewer/index.html [hunk 14]
```diff
@@ -33,6 +34,25 @@
...
+  right: 0;
+  margin: 5px 10px 0 0;
+  border: none;
+  padding: 5px 5px;
+  cursor: pointer;
+}
+
+#rendered-md pre:hover .copy{
+	display: block !important;
+} */
```
Comment (laurent22) [inline ElectronClient/app/gui/note-viewer/index.html]: remove

File: ElectronClient/app/gui/note-viewer/index.html [hunk 15]
```diff
@@ -43,6 +63,28 @@
...
+	// 				setTimeout(() => button.innerText = '', 1000);
+	// 			 } catch (e) {
+	// 				console.error(e);
+	// 			 }
+	// 		}
+	// 	}
+	// })
+
+
+
```
Comment (laurent22) [inline ElectronClient/app/gui/note-viewer/index.html]: remove

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+}
+.copy-code-wrapper button {
+   background: #fff;
+   border: none;
+   padding: 10px;
+   cursor: pointer;
+}
+.copy-code-wrapper button i {
+font-size: 1rem;
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Please use tabs, not spaces for indentation.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+function strEncode(str = '') {
+	if (!str || str.length === 0) {
+		return '';
+	}
+	return str
+		.replace(/</g, '&lt;')
+		.replace(/>/g, '&gt;')
+		.replace(/'/g, '&apos;')
+		.replace(/"/g, '&quot;');
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Encoding/decoding doesn't work as expected. For example, I have this in my code block: `<b>M&amp;M's</b>` and press Copy, I want to get that exact text back, with no encoding/decoding. But with your current implementation I would get this back `<b>M&M's</b>`

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+                        } catch (e) {
+                            console.error(e);
+                        }
+                    }
+                },
+                async buttonClick(_btn) {
+                    const copyButtonDataSet = _btn && _btn.dataset ? _btn.dataset : {};
+                    const notifyId = copyButtonDataSet.mdicNotifyId;
+                    const copyNotify = document.getElementById(notifyId);
+                    const notifyDelayTime = copyButtonDataSet.mdicNotifyDelay;
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Please use tabs for indentation for all this.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 19]
```diff
@@ -0,0 +1,137 @@
+const copyStyle = `
+.code-n-btn-wrapper {
+  positon: relative;
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: "positon" is not a valid CSS property, and this is why your class "copy-code-wrapper" didn't work and you needed the hacky margins.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+}
+.copy-code-wrapper {
+  display: none;
+  position: absolute;
+  right: 0;
+  margin: 10px 15px 0 0;
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Once " .code-n-btn-wrapper" is fixed, please use top/right to position the button and remove the margin (which has a hidden dependency to the current theme).

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+.u-mdic-copy-notify {
+  display: none;
+  padding: 4px 12px;
+  background: #fff;
+  color: #ccc;
+  font-size:0.9rem;
+  margin: 0 10px;
+  border-radius: 10px;
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Very light gray on white is very hard to read. Please use colour `#777` for instance.

**Comment by laurent22** [comment]
Still a bit more to do but we're nearly there. Also it doesn't work on mobile so please either make it work, or disable it in this context.

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 18 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Will do

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 17 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: I think I’ll have to remove this entire function because I found out markdown-it already handle encoding and decoding

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 21 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Alright

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 16 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Copy that

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 20 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Ok

File: ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js (hunk 19 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/copy_to_clipboard.js]: Thanks for showing me this!

File: ReactNativeClient/lib/joplin-renderer/MdToHtml.js (hunk 13 above)
Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: Will do

**Comment by tessus** [comment]
I can help you fix this PR. But I need to know which files you actually modified.

If you post a list of the files you modified, I can have a look.

**Comment by Runo-saduwa** [comment]
Thanks @tessus , Git stung me badly.

i Modified 
- `joplin\ReactNativeClient\lib\joplin-renderer\MdToHtml\rules\code_to_clipboard.js`
- `joplin\ReactNativeClient\lib\joplin-renderer\MdToHtml.js`

**Comment by tessus** [comment]
ok, done.

Please learn how to use git. Check out topics like:

- rewrite git history
- deleting commits
- rebasing

Those are things that are often needed. The problem is, if you mess things up, you can lose your work, so you should really know how to use git.

**Comment by Runo-saduwa** [comment]
i'm very grateful, i will work on my git skills

**Comment by tessus** [comment]
You are welcome. Now you can address the issues Laurent mentioned.

**Comment by tessus** [comment]
And do not do a git pull in your local branch !!!!!!!!!!!! Otherwise you mess it up again!

**Comment by Runo-saduwa** [comment]
Ok. noted

**Comment by tessus** [comment]
Delete your local branch and check it out from your remote.

**Comment by Runo-saduwa** [comment]
Ok, will do

**Comment by Runo-saduwa** [comment]
I have made the changes Laurent requested for which are:
- Changing some of the indentation to tabs
- changed copy to copy_to_clipboard in MdToHtml file
- The encoding/decoding works fine now. if you put anything like `<b>M&amp;M's</b>` in your block of code, you'll get the same thing back after copying
- corrected the typo `positon` to `position`
- removed the hacky margins and simply used `top` and 'right` for postioning
- changed grey color of notification to `#777`

Thats about it.

**Comment by Runo-saduwa** [comment]
It seems like due to the fact that React Native has some  designated API for copy and pasting, i might not be able to use this present code to make copying to work on mobile, therefore we might just resolve to disabling the button in mobile but i dont really know how to go about it.

Please i'll love to hear your idea on how to go about this

File: ReactNativeClient/lib/joplin-renderer/MdToHtml.js [hunk 22]
```diff
@@ -200,15 +200,14 @@ class MdToHtml {
...
+		markdownIt.use(rules.copy_to_clipboard(context));
 		markdownIt.use(rules.image(context, ruleOptions));
 		markdownIt.use(rules.checkbox(context, ruleOptions));
 		markdownIt.use(rules.link_open(context, ruleOptions));
 		markdownIt.use(rules.html_image(context, ruleOptions));
 		if (this.pluginEnabled('katex')) markdownIt.use(rules.katex(context, ruleOptions));
 		if (this.pluginEnabled('fountain')) markdownIt.use(rules.fountain(context, ruleOptions));
 		if (this.pluginEnabled('mermaid')) markdownIt.use(rules.mermaid(context, ruleOptions));
-		markdownIt.use(rules.sanitize_html(context, ruleOptions));
 		markdownIt.use(rules.highlight_keywords(context, ruleOptions));
```
Comment (tessus) [inline ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: Why is the `sanitize_html` deleted? Please add it back. That's probably the reason why the tests fail.

Comment (Runo-saduwa) [reply ReactNativeClient/lib/joplin-renderer/MdToHtml.js]: I don’t know how that happened but I’ll add it back anyway

**Comment by tessus** [comment]
> It seems like due to the fact that React Native has some designated API for copy and pasting, i might not be able to use this present code to make copying to work on mobile, therefore we might just resolve to disabling the button in mobile but i dont really know how to go about it.

I don't underatand what you are saying. Why should this not work on mobile? Did you test it?

**Comment by Runo-saduwa** [comment]
When Laurent reviewed this PR, he said it doesn’t work for mobile, then he said I should either make it work or disable it based on that context. 

I haven’t been able to build the mobile app on my machine because I don’t have Android studio, so I can’t test for mobile. 

So in conclusion, I’m trying to disable this plugin for mobile

**Comment by Runo-saduwa** [comment]
Is there a way to build the mobile app using expo?, because that’s my last resort atm

**Comment by tessus** [comment]
No idea, I just installed Android Studio and went from there.

You still haven't added thta one line. How long does it take to add one line?
We need to see, if this fixes the test error.

**Comment by tessus** [comment]
We are reaching almost 80 comments. Not even a bug report had these many comments so far.
Maybe we should just forget about this and you open a new PR, when you are done with the rest. Don't get me wrong, but you should figure the last part out by yourself. Read some documentation, post in forums, or whatnot. If we tell you what to do all the time, we don't need the PR. How do you ever expect to work on a GSoC project that spans 2 months?

**Comment by Runo-saduwa** [comment]
@tessus You’ve been of great help, I will take your advice and simply close this PR and come back stronger with a complete version. I will deliver
